File: src/camera/discovery.py
"""Camera device discovery and probing."""

from __future__ import annotations

import glob
import sys
from typing import Any

import cv2
import numpy as np

from src.utils.opencv import suppress_opencv_logs, v4l2_backend


def create_capture(device: int | str) -> cv2.VideoCapture:
    return cv2.VideoCapture(device, v4l2_backend())


def probe_capture(device: int | str) -> cv2.VideoCapture | None:
    """Open a device and verify at least one frame can be read."""
    cap = create_capture(device)
    if not cap.isOpened():
        cap.release()
        return None

    ok, frame = cap.read()
    if not ok or frame is None:
        cap.release()
        return None

    return cap


def build_device_candidates(
    device_path: str | None,
    device_id: int,
    max_index: int = 10,
) -> list[int | str]:
    candidates: list[int | str] = []

    if device_path:
        candidates.append(device_path)

    if device_id not in candidates:
        candidates.append(device_id)

    if sys.platform.startswith("linux"):
        for path in sorted(glob.glob("/dev/video*")):
            if path not in candidates:
                candidates.append(path)

    for index in range(max_index):
        if index not in candidates:
            candidates.append(index)

    return candidates


def discover_capture(
    device_path: str | None,
    device_id: int,
) -> tuple[int | str, cv2.VideoCapture] | None:
    """Return the first working device and an already-open capture handle."""
    with suppress_opencv_logs():
        for device in build_device_candidates(device_path, device_id):
            cap = probe_capture(device)
            if cap is not None:
                return device, cap
    return None


def list_devices(max_index: int = 10) -> list[dict[str, Any]]:
    """Return capture devices that OpenCV can open and read from."""
    devices: list[dict[str, Any]] = []
    seen: set[int | str] = set()

    def record(device: int | str, frame: np.ndarray) -> None:
        if device in seen:
            return
        seen.add(device)
        height, width = frame.shape[:2]
        path = device if isinstance(device, str) else f"index {device}"
        devices.append(
            {
                "device": device,
                "path": path,
                "resolution": f"{width}x{height}",
                "works": True,
            }
        )

    candidates: list[int | str] = []
    if sys.platform.startswith("linux"):
        candidates.extend(sorted(glob.glob("/dev/video*")))
    candidates.extend(range(max_index))

    with suppress_opencv_logs():
        for device in candidates:
            cap = create_capture(device)
            if not cap.isOpened():
                cap.release()
                continue
            ok, frame = cap.read()
            if ok and frame is not None:
                record(device, frame)
            cap.release()

    return devices

File: src/camera/discovery.cpp
#include "chess_robot/camera/discovery.hpp"

#include <cstdlib>
#include <filesystem>
#include <set>
#include <sstream>

#include <opencv2/core/utils/logger.hpp>

namespace chess_robot {

namespace fs = std::filesystem;

int videoBackend() {
#ifdef __linux__
  return cv::CAP_V4L2;
#else
  return cv::CAP_ANY;
#endif
}

cv::VideoCapture createCapture(const std::string& device) {
  return cv::VideoCapture(device, videoBackend());
}

cv::VideoCapture createCapture(int device_index) {
  return cv::VideoCapture(device_index, videoBackend());
}

std::optional<cv::VideoCapture> probeCapture(const std::string& device) {
  cv::VideoCapture cap = createCapture(device);
  if (!cap.isOpened()) {
    return std::nullopt;
  }
  cv::Mat frame;
  if (!cap.read(frame) || frame.empty()) {
    cap.release();
    return std::nullopt;
  }
  return cap;
}

std::optional<cv::VideoCapture> probeCapture(int device_index) {
  cv::VideoCapture cap = createCapture(device_index);
  if (!cap.isOpened()) {
    return std::nullopt;
  }
  cv::Mat frame;
  if (!cap.read(frame) || frame.empty()) {
    cap.release();
    return std::nullopt;
  }
  return cap;
}

std::vector<std::string> buildDeviceCandidates(const std::optional<std::string>& device_path,
                                               int device_id,
                                               int max_index) {
  std::vector<std::string> candidates;
  std::set<std::string> seen;

  auto add = [&](const std::string& value) {
    if (seen.insert(value).second) {
      candidates.push_back(value);
    }
  };

  if (device_path.has_value() && !device_path->empty()) {
    add(*device_path);
  }
  add(std::to_string(device_id));

#ifdef __linux__
  for (int i = 0; i < 64; ++i) {
    const fs::path path = fs::path("/dev") / ("video" + std::to_string(i));
    if (fs::exists(path)) {
      add(path.string());
    }
  }
#endif

  for (int index = 0; index < max_index; ++index) {
    add(std::to_string(index));
  }
  return candidates;
}

namespace {

bool isIntegerToken(const std::string& token, int* out) {
  if (token.empty()) {
    return false;
  }
  char* end = nullptr;
  const long value = std::strtol(token.c_str(), &end, 10);
  if (end == token.c_str() || *end != '\0') {
    return false;
  }
  *out = static_cast<int>(value);
  return true;
}

std::optional<cv::VideoCapture> probeToken(const std::string& token) {
  int index = 0;
  if (isIntegerToken(token, &index) && token.find('/') == std::string::npos) {
    return probeCapture(index);
  }
  return probeCapture(token);
}

cv::VideoCapture openToken(const std::string& token) {
  int index = 0;
  if (isIntegerToken(token, &index) && token.find('/') == std::string::npos) {
    return createCapture(index);
  }
  return createCapture(token);
}

}  // namespace

std::optional<DiscoveredCapture> discoverCapture(const std::optional<std::string>& device_path,
                                                 int device_id) {
  const auto previous = cv::utils::logging::getLogLevel();
  cv::utils::logging::setLogLevel(cv::utils::logging::LOG_LEVEL_SILENT);

  std::optional<DiscoveredCapture> result;
  for (const std::string& device : buildDeviceCandidates(device_path, device_id)) {
    auto cap = probeToken(device);
    if (cap.has_value()) {
      result = DiscoveredCapture{device, std::move(*cap)};
      break;
    }
  }

  cv::utils::logging::setLogLevel(previous);
  return result;
}

std::vector<DeviceInfo> listDevices(int max_index) {
  std::vector<DeviceInfo> devices;
  std::set<std::string> seen;

  auto record = [&](const std::string& device, const cv::Mat& frame) {
    if (!seen.insert(device).second) {
      return;
    }
    std::ostringstream resolution;
    resolution << frame.cols << "x" << frame.rows;
    const bool is_index = device.find('/') == std::string::npos;
    devices.push_back(DeviceInfo{
        is_index ? ("index " + device) : device,
        resolution.str(),
        true,
    });
  };

  std::vector<std::string> candidates;
#ifdef __linux__
  for (int i = 0; i < 64; ++i) {
    const fs::path path = fs::path("/dev") / ("video" + std::to_string(i));
    if (fs::exists(path)) {
      candidates.push_back(path.string());
    }
  }
#endif
  for (int index = 0; index < max_index; ++index) {
    candidates.push_back(std::to_string(index));
  }

  const auto previous = cv::utils::logging::getLogLevel();
  cv::utils::logging::setLogLevel(cv::utils::logging::LOG_LEVEL_SILENT);

  for (const std::string& device : candidates) {
    cv::VideoCapture cap = openToken(device);
    if (!cap.isOpened()) {
      continue;
    }
    cv::Mat frame;
    if (cap.read(frame) && !frame.empty()) {
      record(device, frame);
    }
    cap.release();
  }

  cv::utils::logging::setLogLevel(previous);
  return devices;
}

}  // namespace chess_robot
